ceph-mon/lib/charms/ceph_mon/v0/ceph_cos_agent.py

```diff
@@ -45,25 +45,41 @@ def __init__(self, charm):
 
     def _on_refresh(self, event):
         """Enable prometheus on relation change"""
-        if not ceph_utils.is_bootstrapped():
-            logger.debug("not bootstrapped, defer _on_refresh: %s", event)
-            event.defer()
+        if not self._charm.unit.is_leader():
+            logger.debug("Not the charm leader, skipping refresh cb.")
             return
-        logger.debug("refreshing cos_agent relation")
-        if self._charm.unit.is_leader():
-            self.mgr_config_set_rbd_stats_pools()
+
+        if hasattr(self._charm, "_cos_agent_refresh_cb") and callable(self._charm._cos_agent_refresh_cb):
+            self._charm._cos_agent_refresh_cb(event) 
```

Comment:
A bit uneasy about this, probing for a specific method in the client class smells of overly tight coupling. I wonder if we could inject this dependency in the constructor as an optional callback and just call it here?

Reply:
you are right, changing this to constructor passed methods right away.